kinesumer.go

```diff
@@ -626,6 +668,103 @@ func (k *Kinesumer) getNextShardIterator(
 	return output.ShardIterator, nil
 }
 
+/*
+
+Sequence Number management
+
+*/
+func (k *Kinesumer) commitPeriodically() {
+	var checkPointTicker = time.NewTicker(k.commitInterval)
+
+	for {
+		select {
+		case <-k.stop:
+			// If stop channel is closed, consumer goroutine must call Auto function.
+			// Because commitPeriodically function doesn't know whether remained events from Kinesis exist.
+			return
+		case <-k.close:
+			// If close channel is closed, consumer goroutine must call Auto function.
+			// Because commitPeriodically function doesn't know whether remained events from Kinesis exist.
+			return
+		case <-checkPointTicker.C:
+			k.Commit()
+		}
+	}
+}
+
+// MarkRecord marks the provided record as consumed.
+func (k *Kinesumer) MarkRecord(record *Record) {
+	if record == nil {
+		k.sendOrDiscardError(errMarkNilRecord)
+		return
+	}
+
+	seqNum := *record.SequenceNumber
+	if seqNum == "" {
+		// sequence number can't be empty.
+		k.sendOrDiscardError(ErrEmptySequenceNumber)
+		return
+	}
+	if _, found := k.checkPoints[record.Stream]; !found {
+		k.sendOrDiscardError(ErrInvalidStream)
+		return
+	}
+	k.offsets[record.Stream].Store(record.ShardID, seqNum)
+}
+
+// Commit updates check point using current checkpoints.
+func (k *Kinesumer) Commit() {
+	var wg sync.WaitGroup
+	for stream := range k.shards {
+		wg.Add(1)
+
+		var checkpoints []*shardCheckPoint
+		k.offsets[stream].Range(func(shardID, seqNum interface{}) bool {
+			checkpoints = append(checkpoints, &shardCheckPoint{
+				Stream:         stream,
+				ShardID:        shardID.(string),
+				SequenceNumber: seqNum.(string),
+			})
+			return true
+		})
+
+		go func(stream string, checkpoints []*shardCheckPoint) {
+			defer wg.Done()
+			k.commitCheckPointPerStream(stream, checkpoints)
+		}(stream, checkpoints)
+	}
+	wg.Wait()
+}
+
+// update checkpoint using sequence number.
+func (k *Kinesumer) commitCheckPointPerStream(stream string, checkpoints []*shardCheckPoint) {
+	if len(checkpoints) == 0 {
+		k.sendOrDiscardError(errEmptyCommitCheckpoints)
+		return
+	}
+
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), k.commitTimeout)
+	defer cancel()
+
+	if err := k.stateStore.UpdateCheckPoints(timeoutCtx, checkpoints); err != nil {
+		k.sendOrDiscardError(errors.Wrapf(err, "failed to commit on stream: %s", stream))
+		return
+	}
+	go k.cleanupOffsets(checkpoints)
+}
+
+// cleanupOffsets remove uninterested offset which is not updated.
+// TODO(proost): how to remove unused stream?
+func (k *Kinesumer) cleanupOffsets(checkpoints []*shardCheckPoint) {
+	for _, checkpoint := range checkpoints {
+		currentOffset, _ := k.offsets[checkpoint.Stream].LoadAndDelete(checkpoint.ShardID)
+		if currentOffset == checkpoint.SequenceNumber {
+			continue // That committed offset and current offset are same means shard may not be used.
```

Comment:
Is this always true?

Reply:
No. If user call manually `MarkRecord` before `cleanupOffsets` called, then sequence numbers are different

Reply:
Ah, I was asking if `That committed offset and current offset are same means shard may not be used.` is true.

Reply:
Yes. That is true. What i intend is remove shard which may not be used. So it can reduce memory allocation for offsets

Reply:
[ab0d71e](https://github.com/daangn/kinesumer/pull/12/commits/ab0d71e903df3f5d3038c0388362c845387d55ce)

Reply:
[b289e0d](https://github.com/daangn/kinesumer/pull/12/commits/b289e0dfd155444331c3e98ee398b20bc751af0e)